cpp-ch/local-engine/Parser/InputFileNameParser.cpp

```diff
@@ -0,0 +1,224 @@
+    }
+
+private:
+    std::optional<String> file_name;
+    std::optional<Int64> block_start;
+    std::optional<Int64> block_length;
+};
+
+class InputFileExprProjectStep : public DB::ITransformingStep
+{
+public:
+    InputFileExprProjectStep(
+        const DB::DataStream& input_stream,
+        const std::optional<String>& file_name,
+        const std::optional<Int64>& block_start,
+        const std::optional<Int64>& block_length)
+        : ITransformingStep(input_stream, getOutputHeader(input_stream, file_name, block_start, block_length), getTraits(), true)
+        , file_name(file_name)
+        , block_start(block_start)
+        , block_length(block_length)
+    {
+    }
+
+    String getName() const override { return "InputFileExprProjectStep"; }
+    void transformPipeline(DB::QueryPipelineBuilder & pipeline, const DB::BuildQueryPipelineSettings &  /*settings*/) override
+    {
+        pipeline.addSimpleTransform(
+            [&](const DB::Block & header)
+            {
+                return std::make_shared<InputFileExprProjectTransform>(header, output_stream->header, file_name, block_start, block_length);
+            });
+    }
+
+protected:
+    void updateOutputStream() override
+    {
+        output_stream = createOutputStream(input_streams.front(), output_stream->header, getDataStreamTraits());
+    }
+
+private:
+    std::optional<String> file_name;
+    std::optional<Int64> block_start;
+    std::optional<Int64> block_length;
+};
+
+bool InputFileNameParser::hasInputFileNameColumn(const DB::Block & block)
+{
+    auto names = block.getNames();
+    return std::find(names.begin(), names.end(), INPUT_FILE_NAME) != names.end();
+}
+
+bool InputFileNameParser::hasInputFileBlockStartColumn(const DB::Block & block)
+{
+    auto names = block.getNames();
+    return std::find(names.begin(), names.end(), INPUT_FILE_BLOCK_START) != names.end();
+}
+
+bool InputFileNameParser::hasInputFileBlockLengthColumn(const DB::Block & block)
+{
+    auto names = block.getNames();
+    return std::find(names.begin(), names.end(), INPUT_FILE_BLOCK_LENGTH) != names.end();
+}
+
+void InputFileNameParser::addInputFileColumnsToChunk(
+    const DB::Block & header,
+    DB::Chunk & chunk,
+    const std::optional<String> & file_name,
+    const std::optional<Int64> & block_start,
+    const std::optional<Int64> & block_length)
+{
+    auto output_columns = chunk.getColumns();
+    for (size_t i = 0; i < header.columns(); ++i)
+    {
+        const auto & column = header.getByPosition(i);
+        if (column.name == INPUT_FILE_NAME)
+        {
+            if (!file_name.has_value())
+                throw DB::Exception(DB::ErrorCodes::LOGICAL_ERROR, "Input file name is not set");
+            auto type_string = std::make_shared<DB::DataTypeString>();
+            auto file_name_column = type_string->createColumn();
+            file_name_column->insertMany(file_name.value(), chunk.getNumRows());
```

Comment:
This should be a const column. Insert mutli same values into a new empty column is not efficient